File: scripts/fix_csv_encoding.py
"""
Fetches the original CSV from the blob URL, decodes it as Windows-1252,
joins multiline quoted cells, deduplicates header rows, and writes a
clean UTF-8 CSV with a corrected canonical header row.
"""

import urllib.request
import csv
import io
import os

SOURCE_URL = (
    "https://blobs.vusercontent.net/blob/"
    "ANEXO%20I_LISTADO%20PROVISIONAL%20RESULTADOS%20FASE%201-"
    "I6EZLq6EYXcsqGn24vHSJ0Q8k0ua3n.csv"
)

OUT_PATH = "/vercel/share/v0-project/public/data/resultados-fase1.csv"

# Canonical header matching the 12 columns in the CSV
CANONICAL_HEADER = [
    "IDENTIFICADOR",
    "APELLIDOS Y NOMBRE",
    "CONOCIMIENTOS GENERALES",
    "CONOCIMIENTOS IDIOMA INGLÉS",
    "APTITUDES",
    "PERSONALIDAD",
    "TOTAL FASE 1",
    "ESTADO PROVISIONAL",
    "RANKING",
    "RANKING CONOCIMIENTOS",
    "RANKING INGLÉS",
    "RANKING APTITUD",
]


def main():
    print(f"[v0] Fetching CSV from blob URL...")
    with urllib.request.urlopen(SOURCE_URL) as resp:
        raw_bytes = resp.read()
    print(f"[v0] Downloaded {len(raw_bytes):,} bytes")

    # Decode as Windows-1252 — the encoding of the original PDF-exported CSV
    raw_text = raw_bytes.decode("windows-1252")
    raw_text = raw_text.replace("\r\n", "\n").replace("\r", "\n")
    print(f"[v0] Decoded. Total chars: {len(raw_text):,}")

    # Use Python's csv reader — it handles quoted multiline fields automatically
    reader = csv.reader(io.StringIO(raw_text))

    out_rows = []
    header_written = False
    data_count = 0

    for row in reader:
        if not row or all(c.strip() == "" for c in row):
            continue

        # Normalise each cell: strip whitespace, collapse internal newlines/spaces
        row = [" ".join(c.split()) for c in row]

        # Detect header rows: first cell is IDENTIFICADOR (case-insensitive)
        if row[0].upper().strip() == "IDENTIFICADOR":
            if not header_written:
                out_rows.append(CANONICAL_HEADER)
                header_written = True
            continue  # skip all duplicate headers

        out_rows.append(row)
        data_count += 1

    print(f"[v0] Parsed {data_count:,} data rows + 1 header")

    # Write clean UTF-8 CSV
    os.makedirs(os.path.dirname(OUT_PATH), exist_ok=True)
    with open(OUT_PATH, "w", encoding="utf-8", newline="") as f:
        writer = csv.writer(f, quoting=csv.QUOTE_MINIMAL)
        writer.writerows(out_rows)

    print(f"[v0] Written to {OUT_PATH}")

    # Sanity check
    print("[v0] First 6 rows:")
    for i, row in enumerate(out_rows[:6]):
        print(f"  [{i}] {row}")

    # Check a row with ranking data
    with_ranking = [r for r in out_rows[1:] if len(r) >= 9 and r[8].strip().isdigit()]
    print(f"[v0] Rows with ranking data: {len(with_ranking):,}")
    if with_ranking:
        print(f"  Sample: {with_ranking[0]}")


if __name__ == "__main__":
    main()

File: scripts/fix-csv-encoding.mjs
/**
 * Fetches the original CSV from the source blob URL, decodes it as Windows-1252 / Latin-1,
 * then writes a clean UTF-8 version to public/data/resultados-fase1.csv
 */

import { writeFileSync } from 'fs'
import { fileURLToPath } from 'url'
import path from 'path'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const OUT_PATH = path.join(__dirname, '../public/data/resultados-fase1.csv')

const SOURCE_URL =
  'https://blobs.vusercontent.net/blob/ANEXO%20I_LISTADO%20PROVISIONAL%20RESULTADOS%20FASE%201-I6EZLq6EYXcsqGn24vHSJ0Q8k0ua3n.csv'

async function main() {
  console.log('[v0] Fetching original CSV from blob URL...')
  const res = await fetch(SOURCE_URL)
  if (!res.ok) throw new Error(`HTTP ${res.status} fetching CSV`)

  const arrayBuffer = await res.arrayBuffer()
  const bytes = new Uint8Array(arrayBuffer)
  console.log(`[v0] Downloaded ${bytes.length} bytes`)

  // Decode using Windows-1252 (superset of Latin-1 that covers the special chars in this file)
  const decoder = new TextDecoder('windows-1252')
  const raw = decoder.decode(bytes)
  console.log('[v0] Decoded as windows-1252')

  // The CSV has multiline quoted headers repeated many times (one per page from the original PDF export).
  // We need to:
  // 1. Re-join the multiline quoted header cells (e.g. "CONOCIMIENTOS\nGENERALES" → "CONOCIMIENTOS GENERALES")
  // 2. Keep only ONE header row at the top
  // 3. Remove all duplicate header rows
  // 4. Write clean UTF-8 CSV

  // First, collapse all CRLF to LF
  let text = raw.replace(/\r\n/g, '\n').replace(/\r/g, '\n')

  // The headers span 2 physical lines because of quoted newlines:
  //   "CONOCIMIENTOS\nGENERALES" and "CONOCIMIENTOS\nIDIOMA INGLÉS"
  // Collapse quoted multi-line fields: if a line starts inside a quote, merge with previous
  const physicalLines = text.split('\n')
  const logicalLines: string[] = []
  let buffer = ''
  let openQuotes = 0

  for (const line of physicalLines) {
    // Count unescaped quotes
    for (let i = 0; i < line.length; i++) {
      if (line[i] === '"') {
        if (line[i + 1] === '"') { i++; continue } // escaped quote
        openQuotes++
      }
    }
    if (buffer === '') {
      buffer = line
    } else {
      buffer += ' ' + line
    }
    if (openQuotes % 2 === 0) {
      logicalLines.push(buffer)
      buffer = ''
    }
  }
  if (buffer) logicalLines.push(buffer)

  // Normalise the canonical header (remove internal spaces from multiline collapse)
  const CANONICAL_HEADER =
    'IDENTIFICADOR,APELLIDOS Y NOMBRE,CONOCIMIENTOS GENERALES,CONOCIMIENTOS IDIOMA INGLÉS,APTITUDES,PERSONALIDAD,TOTAL FASE 1,ESTADO PROVISIONAL,ranking,ranking conocimientos,rank ingles,rank aptitud'

  // Identify header rows — they start with IDENTIFICADOR
  const HEADER_RE = /^IDENTIFICADOR/i

  let headerWritten = false
  const cleanLines: string[] = []

  for (const line of logicalLines) {
    const trimmed = line.trim()
    if (!trimmed) continue

    if (HEADER_RE.test(trimmed)) {
      if (!headerWritten) {
        cleanLines.push(CANONICAL_HEADER)
        headerWritten = true
      }
      // skip all subsequent duplicate headers
      continue
    }

    cleanLines.push(trimmed)
  }

  const output = cleanLines.join('\n') + '\n'
  writeFileSync(OUT_PATH, output, 'utf-8')
  console.log(`[v0] Written ${cleanLines.length} lines (${output.length} chars) to ${OUT_PATH}`)

  // Quick sanity check: show first 5 data rows
  console.log('[v0] First 5 lines:')
  cleanLines.slice(0, 6).forEach((l, i) => console.log(`  [${i}] ${l.slice(0, 120)}`))
}

main().catch((e) => { console.error(e); process.exit(1) })
